In [2]:
from sklearn.datasets import fetch_openml

from sklearn.model_selection import train_test_split
from sklearn.utils import check_random_state
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier

import numpy as np

X, y = fetch_openml("mnist_784", version=1, return_X_y=True, as_frame=False)

In [10]:
val_samples = y.shape[0] * 0.1
train_test_samples = y.shape[0] * 0.9
kf = KFold(n_splits=4)

random_state = check_random_state(0)
permutation = random_state.permutation(X.shape[0])
X = X[permutation]
y = y[permutation]
X = X.reshape((X.shape[0], -1))

X_train_test, X_val, y_train_test, y_val = train_test_split( X, y, train_size=int(train_test_samples), test_size=int(val_samples))

names = [
    "Logistic Regression",
    "Nearest Neighbors",
    "Linear SVM",
    "Decision Tree",
]

classifiers = [
    LogisticRegression(C=50.0 / train_test_samples, penalty="l1", solver="saga", tol=0.1),
    KNeighborsClassifier(3),
    SVC(kernel="linear", C=0.025),
    DecisionTreeClassifier(max_depth=5),
]

scaler = StandardScaler()

print("Train_test: ", train_test_samples)
print("Val: ", val_samples)

Train_test:  63000.0
Val:  7000.0


2 класса
линейный свм (без кернел)
решающие признаки на дереве вывод
макс качество свм 0,95+

l1 and l2
опорные вектора svm
task найти опорные вектора по обуч. свм и выборке

In [11]:
for i, clf in enumerate(classifiers): #l1 and l2
    accuracys = []
    for train_index, test_index in kf.split(X_train_test):
        X_train, X_test = X_train_test[train_index], X_train_test[test_index]
        y_train, y_test = y_train_test[train_index], y_train_test[test_index]
        X_train = scaler.fit_transform(X_train)
        X_test = scaler.transform(X_test)

        clf.random_state = 42
        clf.fit(X_train, y_train)
        accuracys.append(accuracy_score(y_test, clf.predict(X_test)))
    print("Test accuracy of ", names[i]," is:  %.3f" % np.mean(np.asarray(accuracys)))

Test accuracy of  Logistic Regression  is:  0.819
Test accuracy of  Nearest Neighbors  is:  0.944
Test accuracy of  Linear SVM  is:  0.939
Test accuracy of  Decision Tree  is:  0.663


In [9]:
print(classifiers[2]._get_coef)

<bound method BaseSVC._get_coef of SVC(C=0.025, kernel='linear')>


Full mnist accuracy
Test accuracy of  Logistic Regression  is:  0.819
Test accuracy of  Nearest Neighbors  is:  0.944
Test accuracy of  Linear SVM  is:  0.939
Test accuracy of  Decision Tree  is:  0.663